### PR TITLE
Add lightweight Python bindings for deck.gl JSON API

### PR DESCRIPTION
**Pull request by ajduberstein**
To test, run `python setup.py test` at the `bindings/python/` directory.

One could imagine that these Python objects would usefully do some input validation.

#### Background

Publishing as a complement to https://github.com/uber/deck.gl/pull/2747.

#### Change List
- Adds a set of small Python objects that can serialize to JSON readable by the deck.gl JSON API.

### CODE REVIEW

**Comment by ibgreen** [comment]
> Adds a set of small Python objects that can serialize to JSON readable by the deck.gl JSON API.

This looks like a fantastic start. Just some initial questions:

* As far as I can tell at this stage the deck.gl JSON API is not used yet.
* Assume this is because it hasn't been published in as a pre-built script? 
* For now, we could add a script that builds it locally here when this module is built.

**Comment by ajduberstein** [comment]
> * As far as I can tell at this stage the deck.gl JSON API is not used yet.

Right now, I've provided some Python library scaffolding and a few simple objects that serialize to JSON that can be processed by the deck.gl JSON API. I have to create a packaged version of the JSON API module, then I'll be able to have this all start embedding in a Jupyter notebook.

**Comment by ajduberstein** [comment]
Added a Python utility that takes deck.gl-compliant JSON and renders it in a Jupyter notebook cell. Still needs to support map tiles, map interaction, and a two-way API.

image

**Comment by heshan0131** [comment]
Have you tested whether the map cell will render when 1) re-open the notebook 2) exporting the entire notebook as a html? From my experience, there is an extra step if to save the widget state for 2) to work

**Comment by ibgreen** [comment]
> Have you tested whether the map cell will render when 1) re-open the notebook 

@ajduberstein In the version I played with, I had to reevaluate all cells manually on reload.

> 2) exporting the entire notebook as a html? From my experience, there is an extra step if to save the widget state for 2) to work

Not sure if this was considered yet, we should add this to our list of requirements and test/prioritize it.

File: bindings/python/deckgl-widget/js/lib/extension-core.js [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+          this.el.appendChild(createDiv('deck-container'))
+                 .appendChild(createDiv('deck-map-wrapper'))
+                 .appendChild(canvasNode);
+          canvasNode.parentNode.insertBefore(mapNode, canvasNode.nextSibling);
+      },
+
+      value_changed: function() {
+        this.json_input = this.model.get('json_input');
+
+        if (!this.deck) {
```
Comment (heshan0131) [inline bindings/python/deckgl-widget/js/lib/extension-core.js]: Should probably move this to a `this.createDeck` method

File: bindings/python/deckgl-widget/js/lib/extension-core.js [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+      },
+
+      render: function() {
+          this.model.on('change:json_input', this.value_changed, this);
+          var mapNode = createMapDiv()
+          var canvasNode = createCanvas('deck-map-container')
+          this.el.appendChild(createDiv('deck-container'))
+                 .appendChild(createDiv('deck-map-wrapper'))
+                 .appendChild(canvasNode);
+          canvasNode.parentNode.insertBefore(mapNode, canvasNode.nextSibling);
```
Comment (heshan0131) [inline bindings/python/deckgl-widget/js/lib/extension-core.js]: You can add logic here to check whether `model.get('json_input')` is truthy, and call `this.createDeck` with the existing value. Doing so will ensure that the map re-rendered when the notebook is reloaded

**Comment by ajduberstein** [comment]
image

Extension works. For now, run `./dev.sh` in `bindings/python/deck-widget`, click `examples/Introduction.ipynb`, and run to see the results. I'll be adding it to `pip`, as well.

File: bindings/python/deck-widget/package.json [hunk 3]
```diff
@@ -0,0 +1,88 @@
+{
+  "name": "deckgl-widget",
```
[the remark this answers is not in the document]
Comment (ajduberstein) [reply bindings/python/deck-widget/package.json]: We are publishing both a JS module and a Python module for now–this is simply a holdover from the [template I chose](https://github.com/jupyter-widgets/widget-ts-cookiecutter) for this project.

File: bindings/python/deck-widget/.npmignore [hunk 4]
```diff
@@ -0,0 +1,7 @@
+.DS_Store
```
[the remark this answers is not in the document]
Comment (ajduberstein) [reply bindings/python/deck-widget/.npmignore]: Addressed in 3688e230

File: bindings/.gitignore [hunk 5]
```diff
@@ -0,0 +1,3 @@
+*.pyc
```
[the remark this answers is not in the document]
Comment (ajduberstein) [reply bindings/.gitignore]: Addressed in 3688e230

File: bindings/python/deck-widget/package.json [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  "version": "0.1.0",
+  "description": "Jupyter widget for rendering deck.gl in a Jupyter notebook",
+  "keywords": [
+    "jupyter",
+    "jupyterlab",
+    "jupyterlab-extension",
+    "widgets"
+  ],
+  "files": [
```
[the remark this answers is not in the document]
Comment (ajduberstein) [reply bindings/python/deck-widget/package.json]: Addressed in 3688e230

File: bindings/python/deck-widget/.gitignore [hunk 7]
```diff
@@ -0,0 +1,156 @@
+# Byte-compiled / optimized / DLL files
```
[the remark this answers is not in the document]
Comment (ajduberstein) [reply bindings/python/deck-widget/.gitignore]: Addressed in 3688e230

File: bindings/python/deck-widget/examples/introduction.ipynb [hunk 8]
```diff
@@ -0,0 +1,113 @@
+{
```
[the remark this answers is not in the document]
Comment (ajduberstein) [reply bindings/python/deck-widget/examples/introduction.ipynb]: Agree on the docs. I've removed the docs scaffold generated by [widget-ts-cookiecutter](https://github.com/jupyter-widgets/widget-ts-cookiecutter) in b7cf15e3d4. I'll add docs when we've settled on an API.

**Comment by ajduberstein** [comment]
To do list:

- [ ] Persist state when the notebook saves
- [ ] Travis CI
- [ ] Documentation (awaiting API review)

**Comment by ibgreen** [comment]
We usually keep the TODO list in the issue description, not in a comment in the PR.

If there is no issue for this PR, please create one.

**Comment by ibgreen** [comment]
> move deckgl widget JS to a separate repo

@ajduberstein Just move it into the `modules` directory. If anything should be moved to another repo it should be the python code, but I think it should all stay here for now.

**Comment by ajduberstein** [comment]
Closing out this PR in favor of releasing two more modular changes in #3013 and #3035. The feature tracker for the Python library is at #2929&#8212;feel free to follow along there.

**Pull request closed without merging**